File: sdks/canary/python/scenarios/s_email_reset.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""S-EMAIL-RESET canary — tests email endpoints through the real HTTP boundary."""

from __future__ import annotations

import sys
import os
import time
import json

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import Runner, Config, config_from_env, raw_do, contains_leaked_internals


def run(run: Runner, cfg: Config) -> None:
    base = cfg.api_url.rstrip("/") + "/api/v1"
    unique = int(time.time())
    email = f"canary-email-{unique}@llmsafespaces.test"
    password = "canary-email-pwd-123456"

    # P1: Register
    reg_body = json.dumps({"username": f"canaryemail{unique}", "email": email, "password": password}).encode()
    status, _ = raw_do("POST", base + "/auth/register", "", reg_body)
    run.assert_(status in (201, 409), f"register: 201 or 409 (got {status})")

    # P2: Login. Retry on 429: sibling scenarios' JWT logins share the
    # per-IP login bucket (10/min); this scenario's contract is the
    # 200-vs-403 email-verification semantics, not rate limiting.
    login_body = json.dumps({"email": email, "password": password}).encode()
    login_status = 0
    for attempt in range(7):
        login_status, _ = raw_do("POST", base + "/auth/login", "", login_body)
        if login_status != 429:
            break
        time.sleep(10)
    if login_status == 200:
        run.ok("login: 200 (noop mode — auto-verified)")
    elif login_status == 403:
        run.ok("login: 403 (SES mode — unverified)")
    else:
        run.fail("login: unexpected status", f"got {login_status}")

    # P3: Password-reset request → 202
    reset_body = json.dumps({"email": email}).encode()
    status, _ = raw_do("POST", base + "/auth/password-reset/request", "", reset_body)
    run.assert_(status == 202, f"password-reset-request: 202 (got {status})")

    # P4: Password-reset request unknown → 202
    status, _ = raw_do("POST", base + "/auth/password-reset/request", "",
                       json.dumps({"email": "nonexistent-canary@llmsafespaces.test"}).encode())
    run.assert_(status == 202, f"password-reset-request-unknown: 202 (got {status})")

    # P5: Password-reset confirm bogus → 404
    status, _ = raw_do("POST", base + "/auth/password-reset/confirm", "",
                       json.dumps({"token": "canary-bogus", "newPassword": "canary-new-pwd-123456"}).encode())
    run.assert_(status == 404, f"password-reset-confirm-bogus: 404 (got {status})")

    # P6: Verify-email bogus → 404
    status, _ = raw_do("POST", base + "/auth/verify-email", "",
                       json.dumps({"token": "canary-bogus"}).encode())
    run.assert_(status == 404, f"verify-email-bogus: 404 (got {status})")

    # P7: Verify-email resend → 202
    status, _ = raw_do("POST", base + "/auth/verify-email/resend", "",
                       json.dumps({"email": email}).encode())
    run.assert_(status == 202, f"verify-email-resend: 202 (got {status})")

    # P8: Resend unknown → 202
    status, _ = raw_do("POST", base + "/auth/verify-email/resend", "",
                       json.dumps({"email": "ghost-canary@nonexistent.invalid"}).encode())
    run.assert_(status == 202, f"verify-email-resend-unknown: 202 (got {status})")

    # P9: No leaked internals in error responses
    _, leak_resp = raw_do("POST", base + "/auth/password-reset/confirm", "",
                          json.dumps({"token": "x", "newPassword": "canary-valid-pwd"}).encode())
    run.assert_(not contains_leaked_internals(leak_resp), "password-reset-confirm: no leaked internals")


if __name__ == "__main__":
    run_inst = Runner("email-reset", "python-sdk")
    cfg = config_from_env()
    run(run_inst, cfg)
    res = run_inst.print()
    if res.failed > 0:
        sys.exit(1)

File: sdks/canary/typescript/scenarios/s-email-reset.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// S-EMAIL-RESET canary — tests email endpoints through the real HTTP boundary

import { Runner, Config, configFromEnv, rawDo, containsLeakedInternals } from '../canary.js';

async function run(run: Runner, cfg: Config): Promise<void> {
  const base = cfg.apiUrl.replace(/\/$/, '') + '/api/v1';
  const unique = Date.now();
  const email = `canary-email-${unique}@llmsafespaces.test`;
  const password = 'canary-email-pwd-123456';

  // P1: Register
  const [regStatus] = await rawDo('POST', base + '/auth/register', '', Buffer.from(JSON.stringify({ username: `canaryemail${unique}`, email, password })));
  run.assert(regStatus === 201 || regStatus === 409, `register: 201 or 409 (got ${regStatus})`, '');

  // P2: Login
  // Retry on 429: sibling scenarios' logins share the per-IP login bucket
  // (10/min); this scenario's contract is the 200-vs-403 verification
  // semantics, not rate limiting (Python-twin parity).
  let loginStatus = 0;
  for (let attempt = 0; attempt < 7; attempt++) {
    [loginStatus] = await rawDo('POST', base + '/auth/login', '', Buffer.from(JSON.stringify({ email, password })));
    if (loginStatus !== 429) break;
    await new Promise((res) => setTimeout(res, 10_000));
  }
  if (loginStatus === 200) {
    run.ok('login: 200 (noop mode — auto-verified)');
  } else if (loginStatus === 403) {
    run.ok('login: 403 (SES mode — unverified)');
  } else {
    run.fail('login: unexpected status', `got ${loginStatus}`);
  }

  // P3: Password-reset request → 202
  const [resetStatus] = await rawDo('POST', base + '/auth/password-reset/request', '', Buffer.from(JSON.stringify({ email })));
  run.assert(resetStatus === 202, `password-reset-request: 202 (got ${resetStatus})`, '');

  // P4: Password-reset request unknown → 202
  const [unknownStatus] = await rawDo('POST', base + '/auth/password-reset/request', '', Buffer.from(JSON.stringify({ email: 'nonexistent-canary@llmsafespaces.test' })));
  run.assert(unknownStatus === 202, `password-reset-request-unknown: 202 (got ${unknownStatus})`, '');

  // P5: Password-reset confirm bogus → 404
  const [confirmStatus] = await rawDo('POST', base + '/auth/password-reset/confirm', '', Buffer.from(JSON.stringify({ token: 'canary-bogus', newPassword: 'canary-new-pwd-123456' })));
  run.assert(confirmStatus === 404, `password-reset-confirm-bogus: 404 (got ${confirmStatus})`, '');

  // P6: Verify-email bogus → 404
  const [verifyStatus] = await rawDo('POST', base + '/auth/verify-email', '', Buffer.from(JSON.stringify({ token: 'canary-bogus' })));
  run.assert(verifyStatus === 404, `verify-email-bogus: 404 (got ${verifyStatus})`, '');

  // P7: Verify-email resend → 202
  const [resendStatus] = await rawDo('POST', base + '/auth/verify-email/resend', '', Buffer.from(JSON.stringify({ email })));
  run.assert(resendStatus === 202, `verify-email-resend: 202 (got ${resendStatus})`, '');

  // P8: Resend unknown → 202
  const [resendUnknownStatus] = await rawDo('POST', base + '/auth/verify-email/resend', '', Buffer.from(JSON.stringify({ email: 'ghost-canary@nonexistent.invalid' })));
  run.assert(resendUnknownStatus === 202, `verify-email-resend-unknown: 202 (got ${resendUnknownStatus})`, '');

  // P9: No leaked internals in error responses
  const [, leakResp] = await rawDo('POST', base + '/auth/password-reset/confirm', '', Buffer.from(JSON.stringify({ token: 'x', newPassword: 'canary-valid-pwd' })));
  run.assert(!containsLeakedInternals(leakResp.toString()), 'password-reset-confirm: no leaked internals', '');
}

async function main() {
  const r = new Runner('email-reset', 'typescript-sdk');
  const cfg = configFromEnv();
  await run(r, cfg);
  const res = r.print();
  if (res.failed > 0) process.exit(1);
}

main().catch((e) => { console.error(e); process.exit(1); });
